docs/execution-providers/EP-Context-Design.md

```diff
@@ -93,4 +93,116 @@ It is hard for Execution Providers to generate the partitioned graph within the
+
+
+# EPContext with Weight Sharing
+
+## Weight Sharing in Onnx Domain
+In ONNX, weight sharing refers to multiple ONNX models with external weights pointing to the same external weight file. These models use the same tensor names, allowing them to reference the same tensor data.
+<p align="center"><img width="50%" src="../../images/Onnx_weight_sharing.png" alt="Weight sharing across Onnx models"/></p>
+
+## Weight Sharing in EP Domain with EPContext
+EP weight sharing is enabled using a pre-generated EP context binary/blob. This requires users to generate the context binary offline (AOT) on a Linux x86_64 or Windows x86_64 machine. The EP context binary contains multiple graphs that share the same tensors.
+<p align="center"><img width="50%" src="../../images/EP_weight_sharing.png" alt="Weight sharing in EP context binary"/></p>
+
+The EP or backend SDK should be capable of converting and compiling the graph as described above.
+- The EP or SDK should identify identical weights from the existing EP context generated by previously compiled graphs.
+- When new graphs are compiled into the EP context, they should reuse existing weights if they are recognized as identical.
+For example, in `ep_ctx.bin`, `tensor1_1` from `ep_graph1` and `tensor2_1` from `ep_graph2` are identical and both point to the same data offset, `tensor_data1`.
+
+## EPContext Model Generation with Weight Sharing Workflow
+<p align="center"><img width="90%" src="../../images/EP_weight_sharing_workflow.png" alt="Weight sharing workflow"/></p>
+
+Each ONNX Runtime session is associated with an ONNX model. Models that share weights are grouped into a model group, while ONNX Runtime sessions with common properties are organized into a session group. ONNX Runtime introduces two session options: `ep.share_ep_contexts` and `ep.stop_share_ep_contexts` to facilitate session grouping.
+- All ONNX Runtime sessions within the session group should have `ep.share_ep_contexts` enabled.
+- The final ONNX Runtime session uses `ep.stop_share_ep_contexts` to indicate that it is the last session in the group.
+Note: A single ONNX model may contain multiple `EPContext` nodes, depending on the graph partitioning result. However, for simplicity, each model is shown with only one `EPcontext` node here.
+
+## Implementation Guidance for EPContext Model Generation with Weight Sharing
+- Shared Workspace Creation:
+<br/>    The first session creates a shared workspace (e.g., EP Singleton) to share resources with other sessions.
+- EP Context Binary File Naming:
+<br/>    The EP context binary file name is determined by the first session and stored in the shared workspace (e.g., EP Singleton) for use across session groups.
```

Comment:
>binary file name [](http://example.com/codeflow?start=24&length=16)

Is it requirement to have single binary file?
I suppose this binary file is handled by EPs? So they can decide to have different binary file for each EP node in _ctx.onnx?

Reply:
Yes, it handled by EP. So it depend on EP to decide whether a single binary file or not.